File: ports/java/src/gui/widgets/BcsButton.java
package gui.widgets;

import core.kernel.BcsEvent;
import java.util.concurrent.locks.ReentrantLock;

public class BcsButton extends BcsWidget {
    private String text = "";
    private Runnable clickListener;
    private final ReentrantLock lock = new ReentrantLock();

    public BcsButton(BcsWidget parent) {
        super(parent);
    }

    public String getText() {
        lock.lock();
        try {
            return text;
        } finally {
            lock.unlock();
        }
    }

    public void setText(String text) {
        lock.lock();
        try {
            this.text = text;
        } finally {
            lock.unlock();
        }
    }

    public void setClickListener(Runnable listener) {
        lock.lock();
        try {
            this.clickListener = listener;
        } finally {
            lock.unlock();
        }
    }

    @Override
    protected void mouseReleaseEvent(BcsEvent e) {
        super.mouseReleaseEvent(e);
        Runnable listener;
        lock.lock();
        try {
            listener = this.clickListener;
        } finally {
            lock.unlock();
        }

        if (listener != null) {
            listener.run();
        }
    }
}

File: ports/csharp/src/gui/widgets/BcsButton.cs
using System;

namespace Bcs.Gui.Widgets
{
    public class BcsButton : BcsWidget
    {
        private string _text = string.Empty;
        private readonly object _lock = new object();

        public event Action Clicked;

        public BcsButton(BcsWidget parent) : base(parent)
        {
        }

        public string Text
        {
            get { lock (_lock) { return _text; } }
            set { lock (_lock) { _text = value; } }
        }

        protected override void MouseReleaseEvent(Core.Kernel.BcsEvent e)
        {
            base.MouseReleaseEvent(e);
            Clicked?.Invoke();
        }
    }
}
